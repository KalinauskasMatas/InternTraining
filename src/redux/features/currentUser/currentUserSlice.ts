import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "../../store";
import { UserState } from "../../interfaces";

const initialState: UserState =
  JSON.parse(localStorage.getItem("currUser") as string) || {};

const currentUserSlice = createSlice({
  name: "currentUser",
  initialState,
  reducers: {
    setUser: (state, action: PayloadAction<UserState>) => {
      state.fname = action.payload.fname;
      state.surname = action.payload.surname;
      state.email = action.payload.email;
      state.password = action.payload.password;
      localStorage.setItem("currUser", JSON.stringify(state));
    },
    clearUser: (state) => {
      localStorage.removeItem("currUser");
      state = { ...initialState };
      console.log(state);
    },
  },
});

export const { setUser, clearUser } = currentUserSlice.actions;

export const selectCurrUser = (state: RootState) => state.currentUser;

export default currentUserSlice.reducer;
