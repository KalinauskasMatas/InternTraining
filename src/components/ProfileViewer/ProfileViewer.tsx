import { UserState } from "../../interfaces";

import "./ProfileViewer.css";
import defaultIcon from "./assets/default-icon.png";
import { useAppDispatch, useAppSelector } from "../../redux/hooks";
import { updateCurrUser } from "../../redux/features/currentUser/currentUserSlice";
import {
  updateUser,
  usersSetEmail,
} from "../../redux/features/registeredUsers/registeredUsersSlice";

const ProfileViewer = (props: { profile: UserState }) => {
  const dispatch = useAppDispatch();
  const currProfile = useAppSelector((state) => state.currentUser);
  const profile = props.profile;

  const handleChangeEmail = () => {
    const newEmail = prompt("Enter new email");
    if (!newEmail || !newEmail.includes(`@`)) {
      alert("Enter correct email address");
      return;
    }

    if (profile.email === currProfile.email) {
      dispatch(updateCurrUser({ ...profile, email: newEmail }));
    }

    dispatch(usersSetEmail({ user: profile, newEmail }));
  };

  const handleResetPassword = () => {
    const newPassword = prompt("Enter new password");
    if (!newPassword || newPassword.length < 8) {
      alert("Password must be at least 8 characters length");
      return;
    }

    if (profile.email === currProfile.email) {
      dispatch(updateCurrUser({ ...profile, password: newPassword }));
    }

    dispatch(updateUser({ ...profile, password: newPassword }));
  };

  return (
    <section className="profile">
      <h2 className="profile-header">Profile</h2>
      <div className="profile-content">
        <div className="profile-info">
          <div className="profile-picture">
            <img src={defaultIcon} alt="Default profile icon" />
          </div>
          <div className="profile-text">
            <p data-testid="profile-name">
              <b>Name:</b> {profile.fname}
            </p>
            <p data-testid="profile-surname">
              <b>Surname:</b> {profile.surname}
            </p>
            <p data-testid="profile-email">
              <b>Email:</b> {profile.email}
            </p>
          </div>
        </div>
        <div className="profile-buttons">
          <button onClick={handleResetPassword}>Reset password</button>
          <button onClick={handleChangeEmail}>Reset email</button>
        </div>
      </div>
    </section>
  );
};

export default ProfileViewer;
