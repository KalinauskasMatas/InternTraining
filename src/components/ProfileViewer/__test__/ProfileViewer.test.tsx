import ProfileViewer from "../ProfileViewer";

import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";

import renderer from "react-test-renderer";

import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import currentUserSlice from "../../../redux/features/currentUser/currentUserSlice";

afterEach(cleanup);

const testUser = {
  _id: "123456",
  fname: "test",
  surname: "user",
  email: "test@user",
  password: "12345678",
  rentMovies: [],
  isAdmin: false,
};

const store = configureStore({
  reducer: {
    currentUser: currentUserSlice,
  },
});

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <ProfileViewer profile={testUser} />
    </Provider>
  );
});

it("renders correctly 1", () => {
  render(
    <Provider store={store}>
      <ProfileViewer profile={testUser} />
    </Provider>
  );
  expect(screen.getByTestId("profile-name")).toHaveTextContent("test");
  expect(screen.getByTestId("profile-surname")).toHaveTextContent("user");
  expect(screen.getByTestId("profile-email")).toHaveTextContent("test@user");
});

it("matches snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <ProfileViewer profile={testUser} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
