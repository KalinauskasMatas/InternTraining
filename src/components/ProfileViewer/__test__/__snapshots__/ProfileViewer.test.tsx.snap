// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section
  className="profile"
>
  <h2
    className="profile-header"
  >
    Profile
  </h2>
  <div
    className="profile-content"
  >
    <div
      className="profile-info"
    >
      <div
        className="profile-picture"
      >
        <img
          alt="Default profile icon"
          src="default-icon.png"
        />
      </div>
      <div
        className="profile-text"
      >
        <p
          data-testid="profile-name"
        >
          <b>
            Name:
          </b>
           
          test
        </p>
        <p
          data-testid="profile-surname"
        >
          <b>
            Surname:
          </b>
           
          user
        </p>
        <p
          data-testid="profile-email"
        >
          <b>
            Email:
          </b>
           
          test@user
        </p>
      </div>
    </div>
    <div
      className="profile-buttons"
    >
      <button
        onClick={[Function]}
      >
        Reset password
      </button>
      <button
        onClick={[Function]}
      >
        Reset email
      </button>
    </div>
  </div>
</section>
`;
